effects/butterfly-gears/butterfly-gears.c

```diff
@@ -0,0 +1,167 @@
+  short angle;
+  short angleDelta;
+  int uvPos; // TODO handle u,v separately
+  int uvDelta;
+} BallT;
+
+typedef struct {
+  CopListT *cp;
+  CopInsT *upperBallCopper;
+  CopInsT *lowerBallCopper;
+} BallCopListT;
+
+static BallCopListT ballCopList1; // TODO use second copper list and double-buffer
+static BallT ball1;
+static BallT ball2;
+
+#if USE_DEBUG_BITMAP
+#include "data/gears_testscreen_debug.c"
+#else
+#include "data/gears_testscreen.c"
+#endif
+#include "data/texture_butterfly.c"
+#include "data/texture_butterfly2.c"
+
+// Create copper writes to color registers, leave out colors needed for sprites
+static void InitCopperListBall(CopListT *cp, int y) {
+  short i;
+
+  for (i=0; i<ROTZOOM_H; i++) {
+    CopWait(cp, y, COPWAIT_X);
+    CopMove16(cp, color[ 3 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 5 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 7 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 9 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[11 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[13 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[15 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[18 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[20 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[23 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[26 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[28 & COLREG_MASK], 0xf00);
+    CopMoveWord(cp, 0x1fe, 0);
+    y += 2;
+    CopWait(cp, y, COPWAIT_X);
+    CopMove16(cp, color[ 2 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 4 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 6 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 8 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[10 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[12 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[14 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[16 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[19 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[22 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[24 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[27 & COLREG_MASK], 0xf00);
+    CopMoveWord(cp, 0x1fe, 0);
+    y += 2;
+  }
+}
+
+static void MakeBallCopperList(BallCopListT *ballCp) {
+  CopListT *cp = NewCopList(INSTRUCTIONS_PER_BALL * 2 + 100);
+  ballCp->cp = cp;
+  CopInit(cp);
+  CopSetupDisplayWindow(cp, MODE_LORES, X(0), Y0, 320, 280);
+  CopMove16(cp, dmacon, DMAF_SETCLR | DMAF_RASTER);
+  CopLoadPal(cp, &testscreen_pal, 0);
+  CopSetupMode(cp, MODE_LORES, testscreen.depth);
+  CopSetupBitplanes(cp, NULL, &testscreen, testscreen.depth);
+  CopSetupBitplaneFetch(cp, MODE_LORES, X(0), testscreen.width);
+
+  ballCp->upperBallCopper = cp->curr;
+  InitCopperListBall(cp, Y0 + 7);
+
+  ballCp->lowerBallCopper = cp->curr;
+  InitCopperListBall(cp, Y0 + 127);
+
+  CopEnd(cp);
+}
+
+static void Init(void) {
+  MakeBallCopperList(&ballCopList1);
+  CopListActivate(ballCopList1.cp);
+  ball1.texture = texture_butterfly;
+  ball1.angleDelta = 27;
+  ball1.uvDelta = 1 << 25;
```

Comment:
I have no idea how to read `uvDelta` parameter here. What does it mean? These values are magic compared to `angleDelta` :)

Reply:
Now handling u/v separately, added a macro to use floats in code (which get converted to the internal 9-fractional-bits representation)

Reply:
It looks much better now :)